### ISSUE TITLE
Async re-rendering before task completion

### ISSUE THREAD

**Issue by foobahh** [reactjs/redux#1583]
Using redux with react-router, react-router-redux, redux-saga & redux-logger here.

I am curious why with async composition, the app re-renders & logs before the action is complete, example: 

```
action @ 23:46:48.726 APP_INIT 
App.js:17 RENDERING  Object {history: Object, location: Object, params: Object, route: Object, routeParams: Object…}
App.js:17 RENDERING  Object {history: Object, location: Object, params: Object, route: Object, routeParams: Object…}actions: Objectchildren: Objecthistory: Objectlocation: Objectparams: Objectroute: ObjectrouteParams: Objectroutes: Array[2]routing: Objectuser: Object__proto__: Object
index.js:150  action @ 23:46:48.754 REQUEST_USER 
App.js:17 RENDERING  Object {history: Object, location: Object, params: Object, route: Object, routeParams: Object…}
index.js:150  action @ 23:46:48.874 RECEIVE_USER 
```
As you can see here, the initial render occurs as expected, after the INIT action is logged.  However, the following two actions resolve before the logging occurs.  I do not think this behavior is isolated to redux-saga, however.  I have also experienced similar results with redux-actions combined with async-promise, as well as redux-thunk composition.

Note: the logger middleware is provided last in the applyMiddleware function.  

Could anyone shed some light on this?  Is this expected behavior?

**Comment by gaearon** [comment]
I'm afraid there is really not enough information here to answer your question. Since it's a usage question and not a bug in Redux, it might be better to ask it on a dedicated forum like StackOverflow. If you do confirm it's a bug, please submit a minimal fiddle reproducing it, preferably without other dependencies. Redux is a small library, and it is a bit frustrating that many issues are filed against it which have nothing do with it per se when the fault is either in one of complementary libraries or in misunderstanding of the usage or intended behavior. In this particular case I'd be happy to take a look but the question is somewhat vague and it's hard to guess what went wrong, if anything, without having an actual runnable project reproducing the issue, as well as specific instructions as to what behavior was expected. It's not clear, for example, where "RENDERING" log occurs.

I am closing for the above reasons. Unfortunately right now it's just not actionable for us, and we can't even help you because there is very little information.

**Comment by foobahh** [comment]
Apologies, was not meaning to frustrate - @gaearon if you want to replicate, just put a console.log in the App.js of your async example.  You will see the example console out whatever you put there prior to the RECEIVE_POSTS action logging. Just hoping to figure out what you expect behavior to be, prior to diving deeper into the source itself.

**Comment by gaearon** [comment]
React may call render at any point in your component lifecycle. There are no guarantees about how many times it should be called, and generally you should not care about this. Does this help?

Sorry that I was a little agitated in the last message.

**Comment by foobahh** [comment]
@gaearon no worries.  yeah, so my concern was primarily around _when_ the component "says" it was updated.  with a synchronous action, redux-logger will print to the console, and then if you are logging anything inside of the app it will then write that to the log (expected behavior)

```
1. redux logger message
2. anything console'd in the app as it re-renders
```

however, with async actions, I have been noticing something that (at least I found) to be slightly unexpected

```
1. anything console'd in the app as it re-renders
2. redux logger message
```

I have been quite busy lately, but will try & get a POC test soon.  It is unclear to me whether or not this would have any downstream effects on the application itself while scaling (probably not) but my guess is probably something to do with redux-logger & less redux-itself.  if that is the case I will move my conversation there!  Redux is great though, thanks for all your hard work.
